arpsync/watcher.go

```diff
@@ -0,0 +1,190 @@
+package arpsync
+
+import (
+	"fmt"
+	"net"
+	"strconv"
+	"time"
+
+	"github.com/Sirupsen/logrus"
+	"github.com/rancher/go-rancher-metadata/metadata"
+	"github.com/vishvananda/netlink"
+)
+
+var (
+	// DefaultSyncInterval specifies the default value for arpsync interval in seconds
+	DefaultSyncInterval = 120
+)
+
+// ARPTableWatcher checks the ARP table periodically for invalid entries
+// and programs the appropriate ones if necessary based on info available
+// from rancher-metadata
+type ARPTableWatcher struct {
+	syncInterval time.Duration
+	mc           metadata.Client
+}
+
+// Watch starts the go routine to periodically check the ARP table
+// for any discrepancies
+func Watch(syncIntervalStr string, mc metadata.Client) error {
+	logrus.Debugf("arpsync: syncIntervalStr: %v", syncIntervalStr)
+
+	syncInterval := DefaultSyncInterval
+	if i, err := strconv.Atoi(syncIntervalStr); err == nil {
+		syncInterval = i
+
+	}
+
+	atw := &ARPTableWatcher{
+		syncInterval: time.Duration(syncInterval) * time.Second,
+		mc:           mc,
+	}
+
+	go atw.syncLoop()
+
+	return nil
+}
+
+func buildContainersMap(containers []metadata.Container, network metadata.Network) (map[string]*metadata.Container, error) {
+	containersMap := make(map[string]*metadata.Container)
+
+	for index, aContainer := range containers {
+		if !(aContainer.PrimaryIp != "" && aContainer.NetworkUUID == network.UUID) {
+			continue
+		}
+		containersMap[aContainer.PrimaryIp] = &containers[index]
+	}
+
+	return containersMap, nil
+}
+
+func (atw *ARPTableWatcher) syncLoop() {
+
+	logrus.Infof("arpsync: starting monitoring every %v seconds", atw.syncInterval)
+	for {
+		time.Sleep(atw.syncInterval)
+		logrus.Debugf("arpsync: time to sync ARP table")
+		err := atw.doSync()
+		if err != nil {
+			logrus.Errorf("arpsync: while syncing, got error: %v", err)
+		}
+	}
+}
+
+func (atw *ARPTableWatcher) doSync() error {
+	logrus.Debugf("arpsync: checking the ARP table")
+	networks, err := atw.mc.GetNetworks()
+	if err != nil {
+		logrus.Errorf("arpsync: error fetching networks from metadata")
```

Comment:
This is being logged from the caller: https://github.com/rancher/plugin-manager/pull/60/files/544292e0c05930b8d4ceea0fc88af2ba15cfea73#diff-17ce03a91e4d814b1c8a5f001a723155R69. Isn't that sufficient?